### ISSUE TITLE
Loss becomes nan after certain epoch.

### ISSUE THREAD

**Issue by ynsa** [lhwcv/mlsd_pytorch#9]
Hi!
Try to train tiny model on wireframe dataset(for faster checking issue with loss, make dataset smaller):
![image](https://user-images.githubusercontent.com/27861214/141143493-33b7b968-f915-4699-9495-67ccb8574d21.png)
After certain step loss returns nan. Debug shows that some mask in `weighted_bce_with_logits` is fully zeros, and division on the `torch.sum(mask)` returns nan.

How to fix?

Will appreciate any help.

**Comment by vipul2296** [comment]
Getting the same issue after few epochs loss is nan . Did anyone found any solution ?

**Comment by harolddu** [comment]
Stucking on the same problem, did you find any solution?

**Comment by vipul2296** [comment]
No I haven't find any solution yet. If you could get together and solve the
problem it would be great for both of us.

**Comment by michelebechini** [comment]
Did anyone solved this bug?

**Comment by michelebechini** [comment]
@lhwcv do you have any possible solution for the problem of nans noticed here?

**Comment by lhwcv** [comment]
> @lhwcv do you have any possible solution for the problem of nans noticed here?

@michelebechini  Hi, Have you tried reduce the learning rate,  If still NaN,  report your software's  info to me maybe(python version, PyTorch ..)?    I'll see if this month have time to reproduce and fix it.

**Comment by 1994sugar** [comment]
same question

**Comment by VEDFU** [comment]
I have the same question on wireframe and customed datasets, even though I have already used the pretrained model and reduced the learning rate.

**Comment by VEDFU** [comment]
> this may be a solution.

**Comment by VEDFU** [comment]
> The error is in the ground truth mask of the line segmentation that becomes all zero after few epochs when it is used to compute the line segmentation loss.
> 
> It is not related to the learning rate, I tried a train by excluding the line segmentation loss from the loop and it seems to work fine.

this may be a solution.

**Comment by huyhoang17** [comment]
hi @michelebechini 

> The error is in the ground truth mask of the line segmentation that becomes all zero after few epochs when it is used to compute the line segmentation loss.
It is not related to the learning rate, I tried a train by excluding the line segmentation loss from the loop and it seems to work fine.

Did you solve this issue? I trained the model on Wireframe dataset, but the result was not good, the `sAP_10` was stuck at ~30 (6x in paper). How can I improve this training process?
Thanks

**Comment by huyhoang17** [comment]
@VEDFU 
Did you reproduce the result report in the paper on the Wireframe dataset?

**Comment by michelebechini** [comment]
> hi @michelebechini
> 
> > The error is in the ground truth mask of the line segmentation that becomes all zero after few epochs when it is used to compute the line segmentation loss.
> > It is not related to the learning rate, I tried a train by excluding the line segmentation loss from the loop and it seems to work fine.
> 
> Did you solve this issue? I trained the model on Wireframe dataset, but the result was not good, the `sAP_10` was stuck at ~30 (6x in paper). How can I improve this training process? Thanks

I didn't solved the issue because also for me by using this PyTorch implementation the results are not as good as in the original paper. The issue is NOT related to the learning rate but it is simply related to an issue in reading the ground truth masks computed. Moreover notice that the last time that I tried, also the Loss functions have some bugs (no matching loss) and this can strongly affect the final sAP_10.

**Comment by kushnir95** [comment]
It looks like I found one of the reasons for this issue.
Let's look at this piece of code in [mlsd_pytorch/data/wireframe_dset.py](https://github.com/lhwcv/mlsd_pytorch/blob/main/mlsd_pytorch/data/wireframe_dset.py):
![image](https://user-images.githubusercontent.com/17236032/177992855-ee42607a-28fb-4d87-b367-5d3b9bdb2dfa.png)
Providing that the input size is 512x512, junction_map and line_map are (256, 256, 1) Numpy arrays. Accordingly, `junction_map[0]` and `line_map[0]` are (256, 1). As Numpy does broadcasting whenever it's necessary and possible, the code in rows 334-335 executes without errors, but maps in `label[14, ...]` and `label[15, ...]` are incorrect.
One of the possible solutions is to change `junction_map[0]` to `junction_map[:, :, 0]` and `line_map[0]` to `line_map[:, :, 0]`.

**Comment by huyhoang17** [comment]
Hi @kushnir95 ,
Thank you so much. It works. The sAP_10 point had increased from 30 to 56 with mlsd_large model.
